### ISSUE TITLE
leavesForLeafSet.first().get('start') returns an incorrect start position for some emoji characters

### ISSUE THREAD

**Issue by joshfarrant** [facebook/draft-js#1636]
In DraftEditorBlock.react.js ([L190](https://github.com/facebook/draft-js/blob/master/src/component/contents/DraftEditorBlock.react.js#L190)), certain strings cause `leavesForLeafSet.first().get('start')` to incorrectly report the start position of the string.

For example, entering any of the emoji digits from 0-7 (0️⃣ - 7️⃣) causes `leavesForLeafSet.first().get('start')` to incorrectly return the start position as 1, rather than 0. Oddly, entering either 8️⃣ or 9️⃣ works as expected and the start position is correctly returned as 0.

This doesn't appear to have any negative effect on a basic `<Editor />`, however it becomes an issue when using the emoji plugin as the `decoratedText` prop that draft-js passes through to it ends up being an emoji modifier character (⃣).

You can see this in action by trying to enter any of the 0️⃣ - 7️⃣ emoji into the first demo (Simple Emoji Example) on the [emoji-plugin docs site](https://www.draft-js-plugins.com/plugin/emoji). ([Related issue in draft-js-plugins repo](https://github.com/draft-js-plugins/draft-js-plugins/issues/991)).

Fixing the inconsistencies with `leavesForLeafSet.first().get('start')` so that it consistently returns the correct start position of all numeric emoji characters would resolve this issue.

Tested with v0.10.5 of draft-js on MacOS 10.13.1

**Comment by darwinCarrillo** [comment]
@joshfarrant I am experiencing the same issue. Would you know if there are any workarounds for this ?

**Comment by joshfarrant** [comment]
No, not that I've found. My 'solution' has been to fork the emoji plugin and filter out the problematic emoji so they can't cause a problem.

That's not really a solution though. I'd like to see someone with a bit more knowledge of the core of draft-js have a proper look into this issue, as it's quite clearly a bug (and an easily reproducible one at that).

Josh Farrant - Senior Developer at Nimvelo

**Comment by voodee** [comment]
any progress?

**Comment by thaylin79** [comment]
Running into this issue currently as well. Any progress on this? What it looks like to me is that Draft is not taking into account the possibility of strings containing astral symbols. Here's a great post on what I'm talking about: https://mathiasbynens.be/notes/javascript-unicode
Basically, for certain unicode elements like various emojis you'll receive a couple of unicode values that combine to give you the emoji or symbol you desire (🦊🐼) for example which would end up giving you extra characters in the length value. In fact, those 2 emojis, if you test their length, give you 4! If this isn't getting accounted for (which it doesn't seem to be), your entity ranges and all sorts of stuff gets out of wack. :/

**Comment by dobesv** [comment]
Do you happen to have a list of problematic emoji?  Could we make a PR to the draft-js-emoji plugin to filter those out by default?

**Comment by malavshah9** [comment]
Any progress here?

**Comment by riddhi-24** [comment]
Any update?
